python/dolfinx/fem/petsc.py

```diff
@@ -1118,3 +1122,141 @@ def interpolation_matrix(space0: _FunctionSpace, space1: _FunctionSpace) -> PETS
+        x1: Vector to assign values to.
+    """
+    try:
+        x1_nest = x1.getNestSubVecs()
+        for _x0, _x1 in zip(x0, x1_nest):
+            with _x1.localForm() as x:
+                x.array_w[:] = _x0
+    except PETSc.Error:
+        with x1.localForm() as _x:
+            try:
+                start = 0
+                for _x0 in x0:
+                    end = start + _x0.shape[0]
+                    _x.array_w[start:end] = _x0
+                    start = end
+            except IndexError:
+                _x.array_w[:] = _x0
+
+
+@assign.register(PETSc.Vec)
+def _(x0: PETSc.Vec, x1: typing.Union[npt.NDArray[np.floating], list[npt.NDArray[np.floating]]]):
+    """Assign PETSc vector ``x0`` values to (blocked) array(s) ``x1``.
+
+    This function performs the reverse of the assigment performed by the
+    version of :func:`.assign(x0: typing.Union[npt.NDArray[np.floating],
+    list[npt.NDArray[np.floating]]], x1: PETSc.Vec)`.
+
+    Args:
+        x0: Vector that will have its values assigned to ``x1``.
+        x1: An array or list of arrays to assign to.
+    """
+    try:
+        x0_nest = x0.getNestSubVecs()
+        for _x0, _x1 in zip(x0_nest, x1):
+            with _x0.localForm() as x:
+                _x1[:] = x.array_r[:]
+    except PETSc.Error:
+        with x0.localForm() as _x0:
+            try:
+                start = 0
+                for _x1 in x1:
+                    end = start + _x1.shape[0]
+                    _x1[:] = _x0.array_r[start:end]
+                    start = end
+            except IndexError:
+                x1[:] = _x0.array_r[:]
+
+
+@functools.singledispatch
+def assign_function(u: typing.Union[_Function, list[_Function]], x: PETSc.Vec):
+    """Assign :class:`.Function` degrees-of-freedom to to a vector.
+
+    Assigns degree-of-freedom values in values of ``u``, which is
+    possibly a collection of ``Functions``s, to ``x``. When ``u`` holds
+    a list of ``Functions``s, degrees-of-freedom for the ``Functions``s
+    in ``u`` are 'stacked' and assigned to ``x``. See :func:`assign` for
+    documentation on how stacked assignement is handled.
+
+    Args:
+        u: Function(s) to assign degree-of-freedom value from.
+        x1: Vector to assign degree-of-freedom values in ``u`` to.
+    """
+    if x.getType() == PETSc.Vec.Type().NEST:
+        assign([v.x.array for v in u], x)
+    else:
+        if isinstance(u, Collection):
+            data0, data1 = [], []
+            for v in u:
+                bs = v.function_space.dofmap.bs
+                n = v.function_space.dofmap.index_map.size_local
+                data0.append(v.x.array[: bs * n])
+                data1.append(v.x.array[bs * n :])
+            assign(data0 + data1, x)
+        else:
+            assign(u.x.array, x)
+
+
+@assign_function.register(PETSc.Vec)
+def _(x: PETSc.Vec, u: typing.Union[_Function, list[_Function]]):
+    """Assign vector entries to :class:`.Function` degrees-of-freedom.
```

Comment:
```suggestion
    """Assign vector entries to :class:`Function` degrees-of-freedom.
```
Should there really be a `.` here?